Remove FunctionMap from MGConstrainedDoFs

masterleinad opened the pull request:
At the moment we are using a FunctionMap for setting the boundary dofs in MGConstrainedDoFs
although only the boundary indicators are used. This PR replaces this FunctionMap by a std::vector
of boundary_indicators. 
Furthermore, non-primitive FiniteElements are allowed to be used, if all its components are selected in the ComponentMask.

bangerth commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -65,7 +65,19 @@ class MGConstrainedDoFs : public Subscriptor
+   */
+  template <int dim, int spacedim>
+  void set_zero_boundary_dofs(const DoFHandler<dim,spacedim> &dof,
+                              const std::vector<types::boundary_id> &boundary_indicators,
```
I would prefer if we would use a std::set for this argument because it avoids the need to search through the vector or document how exactly it needs to be sorted.

masterleinad commented:
@bangerth I agree and changed the type for the boundary_indicators to std::set

kronbichler commented:
I definitely agree on using `std::set` at this point, which is also consistent with the interface we use other constraints such as non-normal fluxes.
(Despite the fact that I my opinion otherwise is that we should rather reduce the use of `std::set<some_small_data_type_such_as_int>` because it is almost always less efficient than a vector solution in the finite element context.)

tjhei commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -65,7 +65,19 @@ class MGConstrainedDoFs : public Subscriptor
+  /**
+   * Fill the internal data structures with information about interface and boundary dofs.
+   *
+   * This function internally calls the initialize() function above and the
```
grammar. Do you mean "then" instead of "the"?

tjhei commented on source/multigrid/mg_tools.cc:
```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
+  {
+    boundary_indices.resize(dof.get_triangulation().n_global_levels());
+
+    // if for whatever reason we were passed an empty vector, return immediately
```
"set"

tjhei commented on source/multigrid/mg_tools.cc:
```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
-    // First, deal with the simpler
-    // case when we have to identify
-    // all boundary dofs
+    // First, deal with the simpler case when we have to identif all boundary dofs
```
identify?

tjhei commented on source/multigrid/mg_tools.cc:
```diff
@@ -1278,131 +1344,69 @@ namespace MGTools
                           }
                       }
                     else
                       for (unsigned int i=0; i<local_dofs.size(); ++i)
```
what does this for loop do here? (or is this a bug in the github diff display?)

tjhei commented:
This looks good, thanks for taking a look at this mess. I am not particularly happy with the design of ``set_zero_boundary_dofs`` (weird name, calls initialize, then some other functions, etc..), but maybe this is for another day. /run-tests

masterleinad commented:
@tjhei In fact, initialize has to be called separately. This is now fixed in the documentation along with the rest of your comments.

bangerth commented on doc/news/changes.h:
```diff
@@ -360,6 +360,13 @@ inconvenience this causes.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> Improved: Replace by a std::vector for initializing
```
replace *what*? :-)

bangerth commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -65,7 +65,22 @@ class MGConstrainedDoFs : public Subscriptor
+   * about Dirichlet boundary dofs.
+   *
+   * The initialize() function has to be called additionally
+   * to set hanging node constraints.
```
additionally -> in addition to

what does "set hanging node constraints" refer to? Is there a function `set_hanging_node_constraints()`? If so, maybe use that name.

I find the sentence a bit too vague because it only says that one has to do other things as well, but does not specify a temporal order. How about something of the form
```
  Before calling this function, you also have to call `initialize()` and `set_hanging_node_constraints()`.
```

masterleinad replied:
At the moment, the only purpose of  `initialize` is to set up the refinement_edge_indices. Therefore there is no separate function `set_hanging_node_constraints`.

bangerth replied:
Oh, I now understand the sentence you originally wrote. How about replacing it 
by "Before calling this function, you also have to call initialize() in order 
to set up hanging node constraints." ?

masterleinad replied:
Yes, something similar is in the updated version.

bangerth commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -65,7 +65,22 @@ class MGConstrainedDoFs : public Subscriptor
+   * to set hanging node constraints.
+   *
+   * This function can be called multiple times to allow to consider
+   * different sets of boundary_indicators for different components.
```
to consider -> considering

bangerth commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -160,6 +175,25 @@ MGConstrainedDoFs::initialize(const DoFHandler<dim,spacedim> &dof,
+  // allocate an IndexSet for each global level. Contents will be
+  // overwritten inside make_boundary_list.
+  const unsigned int n_levels = dof.get_triangulation().n_global_levels();
+  boundary_indices.resize(n_levels);
```
this makes me wonder what happens to previous content of the array. Is it possible that there was previous content but `size() != n_levels`? I would expect that this indicates an invalid use of the object.

masterleinad replied:
`boundary_indices` is cleared in `initialize` and `clear`. Since this is a private variable there is no possibility that it has a size different from zero or n_levels.

bangerth replied:
> |boundary_indices| is cleared in |initialize| and |clear|. Since this is a
> private variable there is no possibility that it has a size different from
> zero or n_levels.

Then maybe asserting that this is so would be the right thing to do. It also 
makes it clearer for readers like me that nothing else can happen.

masterleinad replied:
Fine with me.

bangerth commented on include/deal.II/multigrid/mg_tools.h:
```diff
@@ -198,6 +198,18 @@ namespace MGTools
+  make_boundary_list (const DoFHandler<dim,spacedim>      &mg_dof,
+                      const std::set<types::boundary_id> &boundary_indicators,
+                      std::vector<IndexSet>                 &boundary_indices,
+                      const ComponentMask               &component_mask = ComponentMask());
```
We've started moving away from returning objects through reference arguments, in particular if they are followed by other input arguments (possibly defaulted).

Both `std::vector` and `IndexSet` now have move constructors and operators. How about just doing as the documentation suggests and instead *returning* `boundary_indices`?

masterleinad replied:
If I am not overlooking something, this would prevent me from calling `make_boundary` multiple times for filling `boundary_indices`. 
Do you suggest to have `boundary_indices` as input argument and return it?

bangerth replied:
Good point. Maybe mention in the documentation of this function (or the previous function it references) that previous content of the object is not overwritten, but added to, to allow calling it multiple times?

bangerth commented on include/deal.II/multigrid/mg_tools.h:
```diff
@@ -198,6 +198,18 @@ namespace MGTools
+  template <int dim, int spacedim>
+  void
+  make_boundary_list (const DoFHandler<dim,spacedim>      &mg_dof,
+                      const std::set<types::boundary_id> &boundary_indicators,
```
we've recently had a patch where we uniformized the spelling to `boundary_ids`. Maybe do this here as well with the argument name.

bangerth commented on source/multigrid/mg_tools.cc:
```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
+      {
+        IndexSet::ElementIterator it_element = boundary_indexset[i].begin();
+        for (; it_element != boundary_indexset[i].end(); ++it_element)
+          boundary_indices[i].insert(*it_element);
```
isn't there a way to do more efficiently something of the form
```
  boundary_indices[i].insert (boundary_indexset[i].begin(), boundary_indexset[i].end());
```
or
```
  boundary_indices[i] = std::set<...> (boundary_indexset[i].begin(), boundary_indexset[i].end());
```
?

masterleinad replied:
Sure, I go with the first option.

bangerth commented on source/multigrid/mg_tools.cc:
```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
+    const ComponentMask &)
+  {
+    Assert(false, ExcNotImplemented());
+  }
```
Out of curiosity, why is this function different? The function above should just compile as is, or what am I missing?

masterleinad replied:
Trying to compile the template for `<1,1>` gives:
`const class dealii::DoFAccessor<0, dealii::DoFHandler<1>, false>’ has no member named ‘get_mg_dof_indices’`

bangerth replied:
I see. I added the function in #2853 just now. If you want, wait till that is merged and rebase -- you may then be able to remove the specializations. Or we can take care of that as a follow-up.

bangerth commented on source/multigrid/mg_tools.cc:
```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
+    std::vector<IndexSet> &,
+    const ComponentMask &)
+  {
+    Assert(false, ExcNotImplemented());
```
same here, of course. by the way, if you already specialize the `<1,1>` and `<1,2>` cases, you ought to also specialize the `<1,3>` case.

masterleinad replied:
Yes, the remaining cases compile fine.

bangerth commented on source/multigrid/mg_tools.cc:
```diff
@@ -1278,131 +1344,65 @@ namespace MGTools
+                          for (unsigned int c=0; c<n_components; ++c)
+                            Assert (nonzero_component_array[c] == false || component_mask[c] == true,
+                                    ExcMessage ("You are using a non-primitive FiniteElement "
+                                                "and try to constrain just some of its components!"));
```
yes, nice check!

bangerth commented:
I like it. I'm not hot on the name `set_zero_boundary_dofs`. I think what you're trying to say is something of the form "build the internal data structures for boundary dofs that are constrained to zero". Could we come up with a better name for this?

masterleinad commented:
Yes, that is what I wanted to say. In the end, we are just filling `boundary_indices` and hence setting the boundary dofs on all the levels we want to constrain to zero. That's way I came up with this name.
Of course, I am open for other suggestions.
I could think of calling it `make_zero_boundary_constraints` similar to `DoFTools::make_zero_boundary_constraints`.

tjhei commented on include/deal.II/multigrid/mg_constrained_dofs.h:
```diff
@@ -65,7 +65,22 @@ class MGConstrainedDoFs : public Subscriptor
   void initialize(const DoFHandler<dim,spacedim> &dof,
                   const typename FunctionMap<dim>::type &function_map,
-                  const ComponentMask &component_mask = ComponentMask());
+                  const ComponentMask &component_mask = ComponentMask()) DEAL_II_DEPRECATED;
```
can you add a "@deprecated Use initialize() followed by set_zero_boundary_dofs() instead."?

tjhei commented on include/deal.II/multigrid/mg_tools.h:
```diff
@@ -198,6 +198,18 @@ namespace MGTools
   /**
+  * The same function as above, but return an IndexSet rather than a
+  * std::set<unsigned int> on each level and use a std::set of
+  * boundary_ids as input
```
missing period.

tjhei commented:
Can you please take another look at the minor comments? This looks ready to be merged otherwise.

bangerth commented:
Yes, `make_zero_boundary_constraints` sounds better to me.

tjhei commented:
> Yes, make_zero_boundary_constraints sounds better to me.

agreed.

masterleinad commented:
@tjhei I addressed your comments.
@bangerth The specialization make_boundary_list<1,1> is still in. I would go with including the removal in PR #2853 or as a follow-up PR.

bangerth commented:
Yes, reasonable suggestion.
